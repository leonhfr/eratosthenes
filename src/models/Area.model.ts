// Packages.
import * as AWS from 'aws-sdk';
import * as debug from 'debug';
import { Area } from '@scenicroutes/wittgenstein';
import * as turfHelpers from '@turf/helpers';
import * as turfInvariant from '@turf/invariant';

// Internal.
import { BUCKET_GEOJSON } from '../constants';
import { DynamoDB, S3 } from '../clients';
import * as Types from '../types';

// Code.
const debugError = debug('eratosthenes:error:area');
const debugVerbose = debug('eratosthenes:verbose:area');

export class AreaModel {
  static async list(): Promise<Types.Result<Area, Error> | Error> {
    const request: AWS.DynamoDB.DocumentClient.ScanInput = {
      TableName: 'area',
    };
    debugVerbose(`list: requesting %o`, request);

    const ok: Array<Area> = [];
    const err: Array<Error> = [];

    try {
      const response: AWS.DynamoDB.DocumentClient.ScanOutput = await DynamoDB.scan(
        request
      ).promise();

      debugVerbose(`list: response %o`, response);

      if (!response.Items) {
        return { ok, err };
      }

      for (const item of response.Items) {
        const maybeArea = Area.create(item);

        maybeArea instanceof Error ? err.push(maybeArea) : ok.push(maybeArea);
      }

      return { ok, err };
    } catch (err) {
      debugError(`list failed: %s`, err.message);
      return err;
    }
  }

  static async updateItem(
    id: string,
    lastScheduledAt: number,
    zonesComputed: boolean
  ): Promise<AWS.DynamoDB.AttributeMap | undefined | Error> {
    const request: AWS.DynamoDB.DocumentClient.UpdateItemInput = {
      TableName: 'area',
      Key: { id },
      UpdateExpression: 'set lastScheduledAt = :time, zonesComputed = :flag',
      ExpressionAttributeValues: {
        ':time': { N: lastScheduledAt },
        ':flag': { BOOL: zonesComputed },
      },
      ReturnValues: 'ALL_NEW',
    };
    debugVerbose(`updateItem: input %o`, request);

    try {
      const response: AWS.DynamoDB.DocumentClient.UpdateItemOutput = await DynamoDB.update(
        request
      ).promise();

      debugVerbose(`updateItem: response %o`, response);

      return response.Attributes;
    } catch (err) {
      debugError(`updateItem failed: %s`, err.message);
      return err;
    }
  }

  static async getAreaGeojson(
    file: string
  ): Promise<turfHelpers.Feature<turfHelpers.Polygon> | Error> {
    const request: AWS.S3.GetObjectRequest = {
      Key: `${file}.geo.json`,
      Bucket: BUCKET_GEOJSON,
    };

    debugVerbose(`getAreaGeojson: requesting %o`, request);

    try {
      const response = await S3.getObject(request).promise();
      debugVerbose(`getAreaGeojson: response %o`, response);

      if (!response.Body) {
        throw new Error('response.Body is undefined');
      }

      const areaGeojson = JSON.parse(response.Body.toString());

      turfInvariant.featureOf(
        areaGeojson as turfHelpers.Feature<any>,
        'Polygon',
        'callingFunction'
      );

      return areaGeojson as turfHelpers.Feature<turfHelpers.Polygon>;
    } catch (err) {
      debugError(`getAreaGeojson failed: %s`, err.message);
      return err;
    }
  }
}
