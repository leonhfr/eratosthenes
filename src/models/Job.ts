// Packages.
import * as AWS from 'aws-sdk';
import * as debug from 'debug';
import { Job } from '@scenicroutes/wittgenstein';

// Internal.
import { DynamoDB } from '../clients';
import * as Types from '../types';

// Code.
const debugError = debug('eratosthenes:error:job');
const debugVerbose = debug('eratosthenes:verbose:job');

export class JobModel {
  static async list(
    limit: number = 100
  ): Promise<Types.Result<Job, Error> | Error> {
    const request: AWS.DynamoDB.DocumentClient.ScanInput = {
      TableName: 'job',
      Limit: limit,
    };
    debugVerbose(`list: requesting %o`, request);

    const ok: Array<Job> = [];
    const err: Array<Error> = [];

    try {
      const response: AWS.DynamoDB.DocumentClient.ScanOutput = await DynamoDB.scan(
        request
      ).promise();

      debugVerbose(`list: response %o`, response);

      if (!response.Items) {
        return { ok, err };
      }

      for (const item of response.Items) {
        const maybeJob = Job.create(item);

        maybeJob instanceof Error ? err.push(maybeJob) : ok.push(maybeJob);
      }

      return { ok, err };
    } catch (err) {
      debugError(`list failed: %s`, err.message);
      return err;
    }
  }

  static async put(
    job: Job
  ): Promise<AWS.DynamoDB.AttributeMap | undefined | Error> {
    const zoneMap = JSON.parse(JSON.stringify(job.zone));

    const request: AWS.DynamoDB.DocumentClient.PutItemInput = {
      TableName: 'job',
      Item: {
        id: job.id,
        page: job.page,
        zone: zoneMap,
      },
      ReturnValues: 'ALL_OLD',
    };

    debugVerbose(`put: request %o`, request);

    try {
      const response: AWS.DynamoDB.DocumentClient.PutItemOutput = await DynamoDB.put(
        request
      ).promise();
      debugVerbose(`put: response %o`, response);

      return response.Attributes;
    } catch (err) {
      debugError(`put failed: %s`, err.message);
      return err;
    }
  }

  static async delete(
    id: string
  ): Promise<AWS.DynamoDB.AttributeMap | undefined | Error> {
    const request: AWS.DynamoDB.DocumentClient.DeleteItemInput = {
      TableName: 'job',
      Key: { id },
    };
    debugVerbose(`delete: request %o`, request);

    try {
      const response: AWS.DynamoDB.DocumentClient.DeleteItemOutput = await DynamoDB.delete(
        request
      ).promise();

      debugVerbose(`delete: response %o`, response);

      return response.Attributes;
    } catch (err) {
      debugError(`delete failed: %s`, err.message);
      return err;
    }
  }

  static async publish() {
    // publish job to sqs
  }

  static async consume() {
    // consume job from sqs
  }
}
