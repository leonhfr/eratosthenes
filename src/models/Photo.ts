// Packages.
import * as AWS from 'aws-sdk';
import * as debug from 'debug';
import { Photo } from '@scenicroutes/wittgenstein';

// Internal.
import { DynamoDB } from '../clients';

// Code.
const debugError = debug('eratosthenes:error:area');
const debugVerbose = debug('eratosthenes:verbose:area');

export class PhotoModel {
  static async put(
    photo: Photo
  ): Promise<AWS.DynamoDB.AttributeMap | undefined | Error> {
    const {
      id,
      owner,
      secret,
      server,
      farm,
      title,
      description,
      ownername,
      views,
      tags,
      latitude,
      longitude,
      context,
      zoneId,
      inside,
    } = photo;

    const request: AWS.DynamoDB.DocumentClient.PutItemInput = {
      TableName: 'zone',
      Item: {
        id,
        owner,
        secret,
        server,
        farm,
        title,
        description,
        ownername,
        views,
        tags,
        latitude,
        longitude,
        context,
        zoneId,
        inside,
      },
      ReturnValues: 'ALL_OLD',
    };

    debugVerbose(`put: request %o`, request);

    try {
      const response: AWS.DynamoDB.DocumentClient.PutItemOutput = await DynamoDB.put(
        request
      ).promise();
      debugVerbose(`put: response %o`, response);

      return response.Attributes;
    } catch (err) {
      debugError(`put failed: %s`, err.message);
      return err;
    }
  }
}
