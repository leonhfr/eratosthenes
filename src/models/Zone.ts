// Packages.
import * as AWS from 'aws-sdk';
import * as debug from 'debug';
import { Zone } from '@scenicroutes/wittgenstein';

// Internal.
import { DynamoDB } from '../clients';
// import * as Types from '../types';

// Code.
const debugError = debug('eratosthenes:error:zone');
const debugVerbose = debug('eratosthenes:verbose:zone');

export class ZoneModel {
  static async list() {}

  static async put(
    zone: Zone
  ): Promise<AWS.DynamoDB.AttributeMap | undefined | Error> {
    const request: AWS.DynamoDB.DocumentClient.PutItemInput = {
      TableName: 'area',
      Item: {
        id: { S: zone.id },
        area: { S: zone.area },
        bbox: { S: JSON.stringify(zone.bbox) },
        zone: { S: zone.zone ? JSON.stringify(zone.zone) : '' },
      },
      ReturnValues: 'ALL_OLD',
    };
    debugVerbose(`put: request %o`, request);

    try {
      const response: AWS.DynamoDB.DocumentClient.PutItemOutput = await DynamoDB.put(
        request
      ).promise();
      debugVerbose(`put: response %o`, response);

      return response.Attributes;
    } catch (err) {
      debugError(`put failed: %s`, err.message);
      return err;
    }
  }
}
