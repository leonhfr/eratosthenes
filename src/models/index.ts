export { AreaModel } from './Area';
export { JobModel } from './Job';
export { PhotoModel } from './Photo';
export { ZoneModel } from './Zone';
