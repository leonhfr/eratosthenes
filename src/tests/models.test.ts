import * as models from '../models';

// Code.
describe('models', () => {
  it('should expose the expected entries', () => {
    expect(models).toMatchSnapshot();
  });
});
