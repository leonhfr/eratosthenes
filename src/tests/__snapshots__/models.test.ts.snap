// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`models should expose the expected entries 1`] = `
Object {
  "AreaModel": [Function],
  "JobModel": [Function],
  "PhotoModel": [Function],
  "ZoneModel": [Function],
}
`;
